src/log.c

```diff
@@ -0,0 +1,361 @@
+        if (logtype != RTE_LOGTYPE_USER1) {
+            rte_vlog(level, logtype, format, ap);
+            break;
+        }
+
+        cid = rte_lcore_id();
+
+        if (log_stats_info[cid].slow) {
+            /* set log limit rate to 5 sec and keep for 10 mins */
+            if (rte_get_timer_cycles() - log_stats_info[cid].slow_begin > LOG_SLOW_INTERNAL_TIME*rte_get_timer_hz()) {
+                log_stats_info[cid].slow = 0;
+            }
+            if ((rte_get_timer_cycles() - log_stats_info[cid].log_begin) < log_internal*rte_get_timer_hz()) {
+                log_stats_info[cid].missed++;
+                break;
+            }
+            /* just output func and line if log is too fast */
+            len = snprintf(log_buf+off, sizeof(log_buf)-off, "%s:%d\n", func, line);
+            dpvs_async_log(level, logtype, cid, log_buf, len, off);
+            break;
+        }
+
+        len = vsnprintf(log_buf+off, sizeof(log_buf)-off, format, ap);                
+        dpvs_async_log(level, logtype, cid, log_buf, len, off);
+    }while(0);
+    
+    va_end(ap);
+    return 0;
+}
+
+static int log_buf_flush(FILE *f)
+{
+    if (f == NULL ) {
+        w_buf.buf[w_buf.pos] = '\0';
+        syslog(w_buf.level, "%s", w_buf.buf);
+    } else {
+        fwrite(w_buf.buf, w_buf.pos, sizeof(w_buf.buf[0]), f);
+        fflush(f);
+    }
+    w_buf.pos = 0;
+    return 0;
+}
+
+static int log_buf_timeout_flush(FILE *f, int timeout)
+{
+    uint64_t now;
+    
+    now = rte_get_timer_cycles();
+    
+    if (w_buf.pos && ((now - w_buf.time) >= timeout*rte_get_timer_hz())) {
+        log_buf_flush(f);
+    }
+    return 0;
+}
+
+static int log_slave_process(void)
+{
+    struct dpvs_log *msg_log;
+    int ret = EDPVS_OK;
+    FILE *f = rte_logs.file;
+
+    /* dequeue LOG from ring, no lock for ring and w_buf */
+    while (0 == rte_ring_dequeue(log_ring, (void **)&msg_log)) { 
+        if (msg_log->log_len > LOG_BUF_MAX_LEN) {
+            dpvs_log_free(msg_log);
+            continue;
+        }
+        if (w_buf.pos + msg_log->log_len >= LOG_BUF_MAX_LEN) {
+            log_buf_flush(f);
+        }
+        
+        if (!w_buf.pos) {
+            w_buf.level = msg_log->log_level - 1;
+            w_buf.time = rte_get_timer_cycles();        
+        }
+        strncpy(w_buf.buf+w_buf.pos, msg_log->data, msg_log->log_len);
+        w_buf.pos += msg_log->log_len;
+
+        log_buf_timeout_flush(f, 5);            
+        dpvs_log_free(msg_log);
+    }
+    
+    log_buf_timeout_flush(f, 5);
+    
+    return ret;
+}
+
+static void log_slave_loop_func(void)
+{
+    g_dpvs_log_thread_ready = 1;
+    while(1){
+        log_slave_process();
+    }
+}
+
+static void log_signal_handler(int signum)
+{
+    if (signum == SIGABRT || signum == SIGSEGV) {
+        printf("\nSignal %d received, preparing to exit...\n",
+                signum);
+    }
+    log_slave_process();
+    log_buf_flush(rte_logs.file); 
+    signal(signum, SIG_DFL);
```

Comment:
DPVS config file uses SIGHUP signal. Does the SIG_DFL affect the config reload ?